### PR TITLE
Start of layer shell implementation

### PR DESCRIPTION
**Pull request by wmww**
Nothing special has been done for layer surfaces yet (they aren't anchored to screen edges, or always on top or bottom), but they do no appear.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
clang says:

/spread/mir/src/server/frontend_wayland/layer_shell_v1.cpp:54:22: error: private field 'surface' is not used [-Werror,-Wunused-private-field]
    WlSurface* const surface;
                     ^
1 error generated.

**Comment by wmww** [comment]
Fixed

**Comment by AlanGriffiths** [comment]
> don't worry, I'm going to sleep very very soon

I thought you were practising CET ready for next week. ;)

File: src/server/frontend_wayland/window_wl_surface_role.cpp [hunk 1]
```diff
@@ -144,12 +144,11 @@ void mf::WindowWlSurfaceRole::initiate_interactive_resize(MirResizeEdge edge)
...
-                session,
-                surface_id(),
-                sink->latest_timestamp_ns(),
-                Shell::UserRequest::resize,
-                edge);
+            shell->request_operation(session,
+                                     surface_id(),
+                                     sink->latest_timestamp_ns(),
+                                     Shell::UserRequest::resize,
+                                     edge);
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/window_wl_surface_role.cpp]: Just an unrelated whitespace change?

File: src/server/frontend_wayland/window_wl_surface_role.h [hunk 2]
```diff
@@ -36,7 +36,7 @@ struct wl_resource;
 
 namespace mir
 {
-    namespace scene
+namespace scene
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/window_wl_surface_role.h]: Just an unrelated whitespace change?

File: src/server/frontend_wayland/window_wl_surface_role.cpp [hunk 3]
```diff
@@ -185,15 +184,19 @@ void mf::WindowWlSurfaceRole::set_max_size(int32_t width, int32_t height)
                 params->max_width.consume();
         }
         else
+        {
             params->max_width = geom::Width{width};
+        }
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/window_wl_surface_role.cpp]: Just an unrelated layout change?

File: src/server/frontend_wayland/window_wl_surface_role.cpp [hunk 4]
```diff
@@ -185,15 +184,19 @@ void mf::WindowWlSurfaceRole::set_max_size(int32_t width, int32_t height)
...
 
         if (height == 0)
         {
             if (params->max_height.is_set())
                 params->max_height.consume();
         }
         else
+        {
             params->max_height = geom::Height{height};
+        }
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/window_wl_surface_role.cpp]: Just an unrelated layout change?

File: src/platform/options/default_configuration.cpp [hunk 5]
```diff
@@ -57,7 +57,7 @@ char const* const mo::composite_delay_opt         = "composite-delay";
 char const* const mo::enable_key_repeat_opt       = "enable-key-repeat";
 char const* const mo::x11_display_opt             = "x11-display-experimental";
 char const* const mo::wayland_extensions_opt      = "wayland_extensions";
-char const* const mo::wayland_extensions_value    = "wl_shell:xdg_wm_base:zxdg_shell_v6";
+char const* const mo::wayland_extensions_value    = "wl_shell:xdg_wm_base:zxdg_shell_v6:layer_shell_v1";
```
Comment (AlanGriffiths) [inline src/platform/options/default_configuration.cpp]: I know I have some guilt in this (I didn't make the way to do it obvious), but we should probably not enable layer_shell_v1 by default.

We need a real implementation (and tests to prove it) before enabling it by default. As it clearly affects window management, we probably need to surface it through miral and let the window manager participate in the decision to enable this.
